docs/Course Wiki/CS Course Offerings/CS374A.md

```diff
@@ -0,0 +1,95 @@
+# CS374A
+
+## Content Covered
+
+### First third (languages)
+
+- Induction, with a focus on application to strings
+- Regular languages
+- Regular expressions
+- Deterministic finite autoamta (DFAs)
+- Nondeterminstic finite automata (NFAs)
+- Regular language transformation via DFA and NFA manipulations
+- Myhill-Nerode theorem and proving irregularity with fooling sets
+- Context free languages and context free grammars (CFGs)
+- Turing machines (only some semesters)
+
+### Middle third (algorithms)
+
+- Divide and conquer algorithms
+- Backtracking algorithms
+- Dynamic programming algorithms
+- Graph algorithms
+- Greedy algorithms
+
+### Final third (complexity)
+
+- Undecidable languages
+- Undecidability of the halting problem
+- NP-Completeness and NP-hardness reductions
+
+## Prerequisites
+
+- CS173
+- CS225
+
+CS374A studies algorithms using theoretical tools covered in CS173 and data structures covered in CS225. For the first third, students should be comfortable with induction and finite state machines (FSMs) from CS173; for the second third, students should be comfortable with data structures from CS225; for the final third, students should be comfortable with proof by contradiction from CS173.
+
+## When to Take it
+
+CS374A is a required course for Computer Science and Computer Engineering majors. It serves as a prerequisite for CS421, a required course for Computer Science majors, and students should complete it by their second-to-last semester. Algorithmic topics covered in CS374, especially dynamic programming, are useful for internship and full-time job interviews, so taking the course earlier may be beneficial.
+
+Many students find CS374a to be a challenging course, so use caution when scheduling it in a semester with many other difficult courses.
+
+## Course Structure
+
+### Lecture
+
+CS374A has lecture on Tuesday and Thursday for 75 mintues. Attendance is optional, and lectures are recorded. Some students prefer to watch Professor Jeff Erickson's CS374 lecture videos instead, which are listed in [Resources](#resources).
+
+### Discussion
+
+CS374A has discussion sections on Wednesday and Friday, where students work with each other and a TA to solve practice questions. While optional, these discussions provide a way to get comfortable with content before homework and exams.
```

Comment:
Mention how discussions usually reinforce lectures and sometimes introduce important topics that lectures won't introduce (language transformations).